triton-constraint-circuit/src/lib.rs

```diff
@@ -1471,4 +1696,328 @@ mod tests {
         let o = ch(0) * z1 - ch(1) * w;
         assert!(!o.find_equivalent_nodes().is_empty());
     }
+
+    #[derive(Debug, Copy, Clone, Eq, PartialEq, test_strategy::Arbitrary)]
+    enum CircuitOperationChoice {
+        Add(usize, usize),
+        Mul(usize, usize),
+    }
+
+    #[derive(Debug, Copy, Clone, Eq, PartialEq, test_strategy::Arbitrary)]
+    enum CircuitInputType {
+        Main,
+        Aux,
+    }
+
+    #[derive(Debug, Copy, Clone, Eq, PartialEq, test_strategy::Arbitrary)]
+    enum CircuitConstantType {
+        Base(#[strategy(arb())] BFieldElement),
+        Extension(#[strategy(arb())] XFieldElement),
+    }
+
+    fn arbitrary_circuit_monad<II: InputIndicator>(
+        num_inputs: usize,
+        num_challenges: usize,
+        num_constants: usize,
+        num_nodes: usize,
+        num_outputs: usize,
+    ) -> BoxedStrategy<Vec<ConstraintCircuitMonad<II>>> {
+        (
+            vec(CircuitInputType::arbitrary(), num_inputs),
+            vec(CircuitConstantType::arbitrary(), num_constants),
+            vec(
+                CircuitOperationChoice::arbitrary(),
+                num_nodes - num_inputs - num_constants - num_outputs,
+            ),
+            vec(arb::<usize>(), num_outputs),
+        )
+            .prop_map(move |(inputs, constants, operations, outputs)| {
+                let builder = ConstraintCircuitBuilder::<II>::new();
+
+                assert_eq!(0, *builder.id_counter.as_ref().borrow());
+                assert!(
+                    builder.all_nodes.borrow().is_empty(),
+                    "fresh hashmap should be empty"
+                );
```

Comment:
Maybe add a comment that these asserts were added because there exists a suspicion that shrinking re-uses objects?